misk/src/main/kotlin/misk/eventrouter/KubernetesClusterConnector.kt

```diff
@@ -0,0 +1,127 @@
+package misk.eventrouter
+
+import com.google.common.reflect.TypeToken
+import io.kubernetes.client.apis.CoreV1Api
+import io.kubernetes.client.models.V1Pod
+import io.kubernetes.client.util.Config
+import io.kubernetes.client.util.Watch
+import misk.logging.getLogger
+import misk.web.actions.WebSocket
+import misk.web.actions.WebSocketListener
+import okhttp3.OkHttpClient
+import okhttp3.Response
+import okio.ByteString
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Singleton
+
+private val logger = getLogger<KubernetesClusterConnector>()
+
+@Singleton
+internal class KubernetesClusterConnector : ClusterConnector {
+  @Inject @ForKubernetesWatching lateinit var executor: ExecutorService
+
+  private lateinit var api: CoreV1Api
+  private val myPodNameSpace: String = System.getenv("MY_POD_NAMESPACE")
+      ?: throw IllegalStateException("MY_POD_NAMESPACE must be set")
+  private val myPodName: String = System.getenv("MY_POD_NAME")
+      ?: throw IllegalStateException("MY_POD_NAME must be set")
+  private val hostMapping = mutableMapOf<String, String>()
+
+  override fun joinCluster(topicPeer: TopicPeer) {
+    val client = Config.defaultClient()
+    client.httpClient.setReadTimeout(0, TimeUnit.MILLISECONDS)
+
+    api = CoreV1Api(client)
+    executor.execute({
+      val watch = Watch.createWatch<V1Pod>(
+          client,
+          api.listNamespacedPodCall(myPodNameSpace, null, null, null, null, null, true, null, null),
+          object : TypeToken<Watch.Response<V1Pod>>() {}.type)
+
+      for (item in watch) {
+        val name = item.`object`.metadata.name
+        val podIP = item.`object`.status.podIP ?: ""
+        when (item.type) {
+          "ADDED" -> hostMapping[name] = podIP
```

Comment:
I don’t think this is safe

Reply:
hostMapping should be immutable, and this function should create a successor hostmapping